In [3]:
import numpy as np
import pandas as pd

In [422]:
csv_df = pd.read_csv(r"C:\Users\123\Downloads\hw_work.csv")
print(csv_df)

     State  Account length  Area code International plan Voice mail plan  \
0       KS             128        415                 No             Yes   
1       OH             107        415                 No             Yes   
2       NJ             137        415                 No              No   
3       OH              84        408                Yes              No   
4       OK              75        415                Yes              No   
...    ...             ...        ...                ...             ...   
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
0                   

In [13]:
print(csv_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   State                   3333 non-null   object 
 1   Account length          3333 non-null   int64  
 2   Area code               3333 non-null   int64  
 3   International plan      3333 non-null   object 
 4   Voice mail plan         3333 non-null   object 
 5   Number vmail messages   3333 non-null   int64  
 6   Total day minutes       3323 non-null   float64
 7   Total day calls         3323 non-null   float64
 8   Total day charge        3315 non-null   float64
 9   Total eve minutes       3324 non-null   float64
 10  Total eve calls         3325 non-null   float64
 11  Total eve charge        3333 non-null   float64
 12  Total night minutes     3333 non-null   float64
 13  Total night calls       3332 non-null   float64
 14  Total night charge      3333 non-null   

## 1.
#### There are 20 columns in the table
#### There are 3333 rows in the table
#### There is missing data in the 'Total day minutes' column - 10 empty cells out of the 3333 in the column.
#### The DF takes 520.9+ KB of memory usage


In [31]:
print(csv_df["Account length"].describe())


count    3333.000000
mean      101.064806
std        39.822106
min         1.000000
25%        74.000000
50%       101.000000
75%       127.000000
max       243.000000
Name: Account length, dtype: float64


## 2.
#### The highest value in the 'Account length' column is 243, while the lowest value is 1. The average value in the column is 101.064806, in a total of 3333 rows.



In [33]:
print(csv_df.describe())

       Account length    Area code  Number vmail messages  Total day minutes  \
count     3333.000000  3333.000000            3333.000000        3323.000000   
mean       101.064806   437.182418               8.099010         179.787150   
std         39.822106    42.371290              13.688365          54.419625   
min          1.000000   408.000000               0.000000           0.000000   
25%         74.000000   408.000000               0.000000         143.700000   
50%        101.000000   415.000000               0.000000         179.400000   
75%        127.000000   510.000000              20.000000         216.500000   
max        243.000000   510.000000              51.000000         350.800000   

       Total day calls  Total day charge  Total eve minutes  Total eve calls  \
count      3323.000000       3315.000000        3324.000000      3325.000000   
mean        100.456214         30.557831         201.033935       100.110677   
std          20.057356          9.25598

In [39]:
print(csv_df.transpose())

                         0      1      2      3      4      5      6     \
State                      KS     OH     NJ     OH     OK     AL     MA   
Account length            128    107    137     84     75    118    121   
Area code                 415    415    415    408    415    510    510   
International plan         No     No     No    Yes    Yes    Yes     No   
Voice mail plan           Yes    Yes     No     No     No     No    Yes   
Number vmail messages      25     26      0      0      0      0     24   
Total day minutes       265.1  161.6  243.4  299.4  166.7  223.4  218.2   
Total day calls         110.0  123.0  114.0   71.0  113.0   98.0   88.0   
Total day charge        45.07  27.47  41.38   50.9  28.34  37.98  37.09   
Total eve minutes       197.4  195.5  121.2   61.9  148.3  220.6  348.5   
Total eve calls          99.0  103.0  110.0   88.0  122.0  101.0  108.0   
Total eve charge        16.78  16.62   10.3   5.26  12.61  18.75  29.62   
Total night minutes     2

In [53]:
print(csv_df.head(5))

  State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1            110.0   
1                     26              161.6            123.0   
2                      0              243.4            114.0   
3                      0              299.4             71.0   
4                      0              166.7            113.0   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4             99.0             16.78   
1             27.47   

In [55]:
print(csv_df.tail(5))

     State  Account length  Area code International plan Voice mail plan  \
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
3328                     36              156.2             77.0   
3329                      0              231.1             57.0   
3330                      0              180.8            109.0   
3331                      0              213.8            105.0   
3332                     25              234.4            113.0   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
3328             26.55              215.5            126.0  

In [230]:
print(csv_df["State"])

0       KS
1       OH
2       NJ
3       OH
4       OK
        ..
3328    AZ
3329    WV
3330    RI
3331    CT
3332    TN
Name: State, Length: 3333, dtype: object


In [81]:
print(csv_df[["State","International plan"]])

     State International plan
0       KS                 No
1       OH                 No
2       NJ                 No
3       OH                Yes
4       OK                Yes
...    ...                ...
3328    AZ                 No
3329    WV                 No
3330    RI                 No
3331    CT                Yes
3332    TN                 No

[3333 rows x 2 columns]


In [146]:
formatted_df = (csv_df[csv_df["International plan"]=="Yes"])
print(formatted_df)

     State  Account length  Area code International plan Voice mail plan  \
3       OH              84        408                Yes              No   
4       OK              75        415                Yes              No   
5       AL             118        510                Yes              No   
7       MO             147        415                Yes              No   
9       WV             141        415                Yes             Yes   
...    ...             ...        ...                ...             ...   
3304    IL              71        510                Yes              No   
3309    VT             100        408                Yes              No   
3317    SD             163        415                Yes              No   
3320    GA             122        510                Yes              No   
3331    CT             184        510                Yes              No   

      Number vmail messages  Total day minutes  Total day calls  \
3                   

In [232]:
formatted_df_2 = csv_df[(csv_df['International plan'] == "Yes") & (csv_df["Churn"] == True)]
print(formatted_df_2)


     State  Account length  Area code International plan Voice mail plan  \
41      MD             135        408                Yes             Yes   
115     ME              36        510                Yes             Yes   
144     VT             117        408                Yes              No   
198     ME             131        510                Yes             Yes   
214     FL              70        510                Yes              No   
...    ...             ...        ...                ...             ...   
3246    NC              77        408                Yes             Yes   
3255    RI             138        510                Yes              No   
3291    MI             119        510                Yes             Yes   
3304    IL              71        510                Yes              No   
3320    GA             122        510                Yes              No   

      Number vmail messages  Total day minutes  Total day calls  \
41                  

In [234]:
print(csv_df.iloc[14])

State                        IA
Account length               62
Area code                   415
International plan           No
Voice mail plan              No
Number vmail messages         0
Total day minutes           NaN
Total day calls             NaN
Total day charge            NaN
Total eve minutes         307.2
Total eve calls            76.0
Total eve charge          26.11
Total night minutes       203.0
Total night calls          99.0
Total night charge         9.14
Total intl minutes         13.1
Total intl calls              6
Total intl charge          3.54
Customer service calls      4.0
Churn                     False
Name: 14, dtype: object


In [164]:
print(csv_df.iloc[14:])

     State  Account length  Area code International plan Voice mail plan  \
14      IA              62        415                 No              No   
15      NY             161        415                 No              No   
16      ID              85        408                 No             Yes   
17      VT              93        510                 No              No   
18      VA              76        510                 No             Yes   
...    ...             ...        ...                ...             ...   
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
14                  

In [338]:
csv_df_index_state = csv_df.set_index('State', inplace=True)


In [340]:
print(csv_df.loc['LA'])

       Account length  Area code International plan Voice mail plan  \
State                                                                 
LA                117        408                 No              No   
LA                172        408                 No              No   
LA                155        415                 No              No   
LA                121        408                 No              No   
LA                151        408                 No              No   
LA                179        408                 No              No   
LA                 99        415                 No              No   
LA                 93        415                 No              No   
LA                149        415                 No             Yes   
LA                 67        510                 No              No   
LA                172        415                 No              No   
LA                 43        415                 No              No   
LA    

In [342]:
csv_df.drop('Area code', axis=1, inplace=True)
print(csv_df)

       Account length International plan Voice mail plan  \
State                                                      
KS                128                 No             Yes   
OH                107                 No             Yes   
NJ                137                 No              No   
OH                 84                Yes              No   
OK                 75                Yes              No   
...               ...                ...             ...   
AZ                192                 No             Yes   
WV                 68                 No              No   
RI                 28                 No              No   
CT                184                Yes              No   
TN                 74                 No             Yes   

       Number vmail messages  Total day minutes  Total day calls  \
State                                                              
KS                        25              265.1            110.0   
OH             

In [370]:
csv_df.reset_index(drop=True, inplace=True)
csv_df.drop(3328, axis=0, inplace=True)
print(csv_df)

      Account length International plan Voice mail plan  \
0                128                 No             Yes   
1                107                 No             Yes   
2                137                 No              No   
3                 84                Yes              No   
4                 75                Yes              No   
...              ...                ...             ...   
3327              79                 No              No   
3329              68                 No              No   
3330              28                 No              No   
3331             184                Yes              No   
3332              74                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
0                        25              265.1            110.0   
1                        26              161.6            123.0   
2                         0              243.4            114.0   
3                      

In [378]:
csv_df['total_min'] = csv_df['Total night minutes']+csv_df['Total intl minutes']
print(csv_df['total_min'])

0       254.7
1       268.1
2       174.8
3       203.5
4       197.0
        ...  
3327    233.2
3329    200.9
3330    206.0
3331    144.2
3332    255.1
Name: total_min, Length: 3332, dtype: float64


In [380]:
print(csv_df['total_min'].mean)

<bound method Series.mean of 0       254.7
1       268.1
2       174.8
3       203.5
4       197.0
        ...  
3327    233.2
3329    200.9
3330    206.0
3331    144.2
3332    255.1
Name: total_min, Length: 3332, dtype: float64>


In [424]:
first_row = csv_df.head(1)
last_row = csv_df.tail(1)
new_df = pd.concat([first_row,last_row],ignore_index=True)
print(new_df)

  State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    TN              74        415                 No             Yes   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1            110.0   
1                     25              234.4            113.0   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4             99.0             16.78   
1             39.85              265.9             82.0             22.60   

   Total night minutes  Total night calls  Total night charge  \
0                244.7               91.0               11.01   
1                241.4               77.0               10.86   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3                2.7   
1                13.7              

In [466]:
new_data_dict = dict(csv_df)
new_data_keys = new_data_dict.keys()
print(new_data_keys)
values = ['AR','299','155','No','Yes','22','214.01','32.51','51.8','411.22','15.3','5.9',',41','23.2','909','404','1920','3.25','2.0','False']


dict_keys(['State', 'Account length', 'Area code', 'International plan', 'Voice mail plan', 'Number vmail messages', 'Total day minutes', 'Total day calls', 'Total day charge', 'Total eve minutes', 'Total eve calls', 'Total eve charge', 'Total night minutes', 'Total night calls', 'Total night charge', 'Total intl minutes', 'Total intl calls', 'Total intl charge', 'Customer service calls', 'Churn'])


In [476]:
new_data = dict(zip(new_data_keys,values))
new_data_row = pd.DataFrame([new_data])
print(new_data_row)

  State Account length Area code International plan Voice mail plan  \
0    AR            299       155                 No             Yes   

  Number vmail messages Total day minutes Total day calls Total day charge  \
0                    22            214.01           32.51             51.8   

  Total eve minutes Total eve calls Total eve charge Total night minutes  \
0            411.22            15.3              5.9                 ,41   

  Total night calls Total night charge Total intl minutes Total intl calls  \
0              23.2                909                404             1920   

  Total intl charge Customer service calls  Churn  
0              3.25                    2.0  False  


In [478]:
new_df = pd.concat([csv_df,new_data_row],ignore_index=True)
print(new_df)

     State Account length Area code International plan Voice mail plan  \
0       KS            128       415                 No             Yes   
1       OH            107       415                 No             Yes   
2       NJ            137       415                 No              No   
3       OH             84       408                Yes              No   
4       OK             75       415                Yes              No   
...    ...            ...       ...                ...             ...   
3329    WV             68       415                 No              No   
3330    RI             28       510                 No              No   
3331    CT            184       510                Yes              No   
3332    TN             74       415                 No             Yes   
3333    AR            299       155                 No             Yes   

     Number vmail messages Total day minutes Total day calls Total day charge  \
0                       25    

In [ ]:
## What does 'inplace' mean?
### The 'inplace' parameter determines whether an operation in the df modifies the original df, or just creats a copy with the changes made. Changes will affect the original if inplace = True, and will not if inplace = False

In [ ]:
## Any column in the table can be set as index, using set_index

In [ ]:
## The axis = 0 refers to rows in the df, while axis = 1 refers to columns. We can use the axis to drop a column or a raw, for example, or to make other changes in the df.

In [ ]:
## In df:
# '&' - Means 'and'
# '|' - Means 'or
# '~' - Means 'not'
# Isin - the'isin' method checks if an element in a column or series is contained in a list of values
# between - the 'between' method checks if an element is bet
~
Isin
between